File: server/src/main/java/org/queue/Queue.scala
/*
 * Copyright 2010 LinkedIn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.queue

import org.queue.consumer.ConsumerConfig
import org.queue.server.{KafkaConfig, KafkaServerStartable}
import org.queue.utils.Utils
import org.slf4j.LoggerFactory

object Queue {
  private val logger = LoggerFactory.getLogger(Queue.getClass)

  def main(args: Array[String]): Unit = {
//    val kafkaLog4jMBeanName = "kafka:type=kafka.KafkaLog4j"
//    Utils.swallow(Level.WARN, Utils.registerMBean(LoggerFactory.getLogger(""), kafkaLog4jMBeanName))

    var serverPath = System.getProperty("server.config")
    if(serverPath == null || serverPath.isBlank) {
      serverPath = getClass().getResource("/server.properties").getPath
    }
    var consumerPath = System.getProperty("consumer.config")
    if(consumerPath == null || serverPath.isBlank) {
      consumerPath = getClass().getResource("/consumer.properties").getPath
    }

    try {
      var kafkaServerStartble: KafkaServerStartable = null
      val props = Utils.loadProps(serverPath)
      val serverConfig = new KafkaConfig(props)
      if (args.length == 2) {
        val consumerConfig = new ConsumerConfig(Utils.loadProps(consumerPath))
        kafkaServerStartble = new KafkaServerStartable(serverConfig, consumerConfig)
      }else {
        kafkaServerStartble = new KafkaServerStartable(serverConfig)
      }

      // attach shutdown handler to catch control-c
      Runtime.getRuntime().addShutdownHook(new Thread() {
        override def run() = {
          kafkaServerStartble.shutdown
          kafkaServerStartble.awaitShutdown
        }
      });

      kafkaServerStartble.startup
      kafkaServerStartble.awaitShutdown
    }catch {
      case e : Throwable => logger.error(e.getMessage,e)
    }
    System.exit(0)
  }
}

File: server/src/main/java/org/queue/Queue.java
package org.queue;

import org.queue.consumer.ConsumerConfig;
import org.queue.server.QueueConfig;
import org.queue.server.QueueServerStartable;
import org.queue.utils.Utils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Properties;

public class Queue {
    private static final Logger logger = LoggerFactory.getLogger(Queue.class);

    public static void main(String[] args) {
        // Utils.swallow(Level.WARN, () -> Utils.registerMBean(LoggerFactory.getLogger(""), queueLog4jMBeanName));

        boolean embeddedConsumer = false;
        String serverPath = System.getProperty("server.config");
        if (serverPath == null || serverPath.isBlank()) {
            // 如果系统属性中没有配置server.config，则使用默认的配置文件路径
            serverPath = Queue.class.getResource("/server.properties").getPath();
        }
        String consumerPath = System.getProperty("consumer.config");
        if (consumerPath == null || consumerPath.isBlank()) {
            // 如果系统属性中没有配置consumer.config，则使用默认的配置文件路径
            consumerPath = Queue.class.getResource("/consumer.properties").getPath();
        }

        try {
            QueueServerStartable queueServerStartable = null;
            Properties props = Utils.loadProps(serverPath);
            QueueConfig serverConfig = new QueueConfig(props);
            if (embeddedConsumer) {
                ConsumerConfig consumerConfig = new ConsumerConfig(Utils.loadProps(consumerPath));
                queueServerStartable = new QueueServerStartable(serverConfig, consumerConfig);
            } else {
                queueServerStartable = new QueueServerStartable(serverConfig);
            }

            // 添加关闭钩子，以便捕获Ctrl-C等中断信号
            QueueServerStartable finalQueueServerStartable = queueServerStartable;
            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
                finalQueueServerStartable.shutdown();
                finalQueueServerStartable.awaitShutdown();
            }));

            queueServerStartable.startup();
            queueServerStartable.awaitShutdown();
        } catch (Throwable e) {
            logger.error("Error occurred while starting queue server", e);
        }
        System.exit(0);
    }
}